src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage.Accessor.cs

```diff
@@ -0,0 +1,158 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.SQLite.Interop;
+using Microsoft.CodeAnalysis.Storage;
+using Roslyn.Utilities;
+using SQLitePCL;
+
+namespace Microsoft.CodeAnalysis.SQLite
+{
+    internal partial class SQLitePersistentStorage
+    {
+        /// <summary>
+        /// Abstracts out access to specific tables in the DB.  This allows us to share overall
+        /// logic around cancellation/pooling/error-handling/etc, while still hitting different
+        /// db tables.
+        /// </summary>
+        private abstract class Accessor<TKey, TWriteQueueKey, TDatabaseId>
+        {
+            protected readonly SQLitePersistentStorage Storage;
+
+            /// <summary>
+            /// Queue of actions we want to perform all at once against the DB in a single transaction.
+            /// </summary>
+            private readonly MultiDictionary<TWriteQueueKey, Action<SqlConnection>> _writeQueue =
+                new MultiDictionary<TWriteQueueKey, Action<SqlConnection>>();
+
+            public Accessor(SQLitePersistentStorage storage)
+            {
+                Storage = storage;
+            }
+
+            protected abstract string DataTableName { get; }
+
+            protected abstract bool TryGetDatabaseId(SqlConnection connection, TKey key, out TDatabaseId dataId);
+            protected abstract void BindFirstParameter(SqlStatement statement, TDatabaseId dataId);
+            protected abstract TWriteQueueKey GetWriteQueueKey(TKey key);
+
+            public Task<Stream> ReadStreamAsync(TKey key, CancellationToken cancellationToken)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (!Storage._shutdownTokenSource.IsCancellationRequested)
+                {
+                    using (var pooledConnection = Storage.GetPooledConnection())
+                    {
+                        var connection = pooledConnection.Connection;
+                        if (TryGetDatabaseId(connection, key, out var dataId))
+                        {
+                            // Ensure all pending document writes to this name are flushed to the DB so that 
+                            // we can find them below.
+                            FlushPendingWrites(connection, key);
```

Comment:
I'm not sure what to do here :)

We could simply say "your read is not guaranteed to see your previous write."  But i thought it would be preferable to give that guarantee.  It also means we would have to have some system for testing by which we could actually verify that reads would eventually see the data.